crates/autopilot/src/on_settlement_event.rs

```diff
@@ -302,3 +202,230 @@ impl Inner {
+    db: Postgres,
+    solvers: Vec<eth::Address>,
+    // Registry to keep track of the already processed settlements, format (block, hash)
+    registry: Arc<Mutex<BTreeMap<i64, H256>>>,
+}
+
+impl CircuitBreaker {
+    /// Maximum capacity of the registry in order not to overflow the memory
+    const MAX_REGISTRY_SIZE: usize = 100;
+    /// The number of settlements taken from the database (ordered by block)
+    const NUMBER_SETTLEMENTS: i64 = 4;
+
+    pub fn build(
+        eth: infra::blockchain::circuit_breaker::CircuitBreaker,
+        persistence: infra::Persistence,
+        solvers: Vec<eth::Address>,
+        db: Postgres,
+    ) -> Arc<dyn OnSettlementEventAction> {
+        Arc::new(Self {
+            circuit_breaker: eth,
+            persistence,
+            db,
+            solvers,
+            registry: Arc::new(Mutex::new(BTreeMap::new())),
+        })
+    }
+}
+
+#[async_trait]
+impl OnSettlementEventAction for CircuitBreaker {
+    async fn apply(&self) -> Result<bool> {
+        let mut ex = self
+            .db
+            .pool
+            .begin()
+            .await
+            .context("acquire DB connection")?;
+        let mut events = database::settlements::get_settlements_without_auction(
+            &mut ex,
+            Self::NUMBER_SETTLEMENTS,
+        )
+        .await
+        .context("get_settlement_without_auction")?
+        .into_iter()
+        .map(|event| (event.block_number, H256(event.tx_hash.0)))
+        .collect::<BTreeMap<_, _>>();
+
+        {
+            let mut registry = self.registry.lock().await;
+
+            let new_events = events
+                .iter()
+                .filter(|&(k, _)| !registry.contains_key(k))
+                .map(|(&k, &v)| (k, v))
+                .collect::<BTreeMap<_, _>>();
+
+            if new_events.is_empty() {
+                return Ok(false);
+            }
+
+            // @TODO: Write here the check for the settlement validity
+
+            // We want to keep the registry with a max size of MAX_REGISTRY_SIZE, since the
+            // BTreeMap is ordered by key, every time there is a buffer
+            // overflow, the oldest blocks are removed This way we always
+            // preserve the most up-to-date data in-memory
+            registry.append(&mut events);
+            if registry.len() > Self::MAX_REGISTRY_SIZE {
+                (0..(registry.len() - Self::MAX_REGISTRY_SIZE)).for_each(|_| {
+                    registry.pop_first();
+                })
+            }
+        }
+
+        Ok(true)
+    }
+}
+
+impl OnSettlementEvent {
+    /// Creates a new OnSettlementEventUpdater
+    pub fn new() -> Self {
+        let inner = Arc::new(Inner {
+            notify: Notify::new(),
+        });
+        Self { inner }
+    }
+
+    /// Asynchronously schedules the first update run.
+    pub fn add_listener(
+        self,
+        on_settlement_event_action: Arc<dyn OnSettlementEventAction>,
```

Comment:
I'm a bit worried about introducing traits here because getting those right can be hard.
What could be done instead to decouple this setup is to have a `broadcast::channel` in the event listener and have a `subscribe` method on it.
That way any component that requires an event stream could just `subscribe` to it (which in turn spawns a new channel they can block on for updates) and each component could do whatever they want with this channel without having to implement any traits.